vumi_http_retry/limiter.py

```diff
@@ -0,0 +1,65 @@
+from twisted.internet.defer import Deferred
+
+
+class DeferredLimiterOverflow(Exception):
+    """Raised when one attempts to add to a full DeferredLimiter"""
+
+
+class DeferredLimiter(object):
+    """
+    Keeps track of a limited list of pending deferreds, allowing one to wait
+    for a vacancy in the list when one of the deferreds is done.
+    """
+    def __init__(self, limit, errback=None):
+        self.limit = limit
+        self.deferreds = []
+        self.observers = []
+
+        if errback is None:
+            errback = lambda f: f
+
+        self.errback = errback
+
+    @property
+    def full(self):
+        return self.count >= self.limit
+
+    @property
+    def count(self):
+        return len(self.deferreds)
+
+    def add(self, d):
+        if self.full:
+            raise DeferredLimiterOverflow()
+
+        self.deferreds.append(d)
+        d.addCallback(self._callback, d)
+        d.addErrback(self._errback, d)
+        return d
+
+    def when_not_full(self):
+        d = Deferred()
```

Comment:
Summary of IRL discussion: Having multiple observers is potentially odd because a deferred returned by `when_not_full` could run after the queue has already filled up again. Given this, it seems sensible to limit the number of observers to one since that matches our use case.